Add CSV import/export for contact blocklist

Raroun opened the pull request:
This adds the ability to import and export blocked contacts as CSV files.

Export: Download your current blocked contacts list as a CSV file (contact URL + block reason).

Import: Upload a CSV file to bulk-block contacts. Shows a preview before importing, with an optional purge checkbox to also remove content from those contacts.

Follows the same pattern as the existing server blocklist import (`/moderation/blocklist/server/import`).

Menu:
![Contact_Blocklist_1](https://github.com/user-attachments/assets/4f3cca9c-6bee-4067-b81a-c8b2066521ea)

Contact Blocklist Menu:
![Contact_Blocklist_2](https://github.com/user-attachments/assets/1a27c9b1-f599-4aa7-b12b-f763d4011bd1)

Preview before importing:
![Contact_Blocklist_3](https://github.com/user-attachments/assets/bc8a30b3-e18d-4a25-9bdc-07693b4c703e)

Summary:
![Contact_Blocklist_4](https://github.com/user-attachments/assets/db6480cc-9915-4b3c-937f-aae8c1b37bad)

Donwload CSV:
![Contact_Blocklist_5](https://github.com/user-attachments/assets/c0d28d02-3571-4eee-885f-51760bfb24e8)

As always, I'm thankfull for any suggestions or feedback.

Closes #15316

Commit:
Add CSV import/export for contact blocklist

Commit:
Update messages.po

Raroun commented:
https://git.friendi.ca/ returns a 502 ( Bad Gateway)

marcusxss commented:
Very useful feature, thanks! Especially with the increase in spam there's been, as far as I've heard.

The purge function seems very useful too.

Does it list individually which ones are already blocked? (It looks like it in the screenshots?)  I imagine such lists may get very long over time, especially as a result of bot accounts, so maybe it should only list the number of entries already on the list, or only show them by name if they're below x amount? (I'd lean towards the former.) 
Fx. maybe friendica.world has a 500+ blocklist, and I import their newest version every couple of months - that will mean a lot of duplicates.

But yes, I also noticed git.friendi.ca is down on one of my PRs.

**General thoughts on block lists, not specifically related to this PR**:

In the future I genuinely wonder whether is a good or bad idea in terms of its consequences, if one could specify that one subscribes to and automatically imports the blocklists from X, Y, Z instance/URL every once in a while. A potentially big downside is that it could be abused to exclude users without good reason very fast from much of the network - the upside is that I imagine as the fediverse grows, spam, bots and bad actors will only become a (much) larger problem, and they will have to be blocked more efficiently to not overrun the network successfully. Of course this PR in itself is something that makes dealing with them a lot easier, so that's great!
Another approach that would likely help is disabling open signups, but maybe that means significantly fewer new, real users in the fediverse.

Raroun commented:
Thanks for the feedback!

Regarding duplicates: Good point, but it actually already limits the output - it only shows the first 10 individual error messages, then just says "... and X more errors". The summary counts ("X contacts were blocked", "Y contacts were skipped") are always shown. So importing a 500+ list with many duplicates should still be manageable.

But if you think even 10 is too many and it should only show the counts without any individual listings, I can change that. :)

marcusxss commented:
I can answer that with Futurama:
![neutral planet](https://github.com/user-attachments/assets/f8d3f9ac-08d1-43d7-a188-e772e675b639)
:-)

Raroun commented:
> In the future I genuinely wonder whether is a good or bad idea in terms of its consequences, if one could specify that one subscribes to and automatically imports the blocklists from X, Y, Z instance/URL every once in a while. A potentially big downside is that it could be abused to exclude users without good reason very fast from much of the network - the upside is that I imagine as the fediverse grows, spam, bots and bad actors will only become a (much) larger problem, and they will have to be blocked more efficiently to not overrun the network successfully. Of course this PR in itself is something that makes dealing with them a lot easier, so that's great!
> Another approach that would likely help is disabling open signups, but maybe that means significantly fewer new, real users in the fediverse.

I totally get where you're coming from—this is a topic that has been debated forever. But honestly, we have to weigh this against server-level blocks. We already see things like #fediblock being used to spread server blocklists, often without much auditing or verification.

Viewed through that lens, importing user-level blocklists is likely the lesser of two evils. It avoids the massive collateral damage of blocking whole instances.

loma-one commented:
Mastodon sends a flag to other Mastodon servers when a spam account is suspended (https://github.com/friendica/friendica/issues/15223). Friendica does not process this flag. As a result, spam messages remain visible in the timeline longer than necessary and can cause damage. After approximately 14 days, the account is deleted. 

My guess is therefore that the list will not be very long due to additional spam accounts. The administrator remains responsible for checking the imported blocks.

annando commented:
@Raroun please have got a look at the failed "Code Quality" check in the pipeline. There seems to be one thing to be fixed.

Concerning the import there is one problem. Currently you process the data while importing. If the list is too long or if a lot of contacts are unknown, then `Contact::getByURL` will take some time - which will result in time out errors, since most web servers are configured to stop processing after some seconds.

Please have got a look at the contact import and see how this is problem is solved there: https://github.com/friendica/friendica/blob/2025.07-rc/src/Module/Settings/ContactImport.php

Raroun commented:
@annando 
Thank you for the feddback - i will check how to create a worker for this to block the contacts asynchronous in the background.

Commit:
Add async CSV import for contact blocklist

Raroun commented:
i changed the adding to blocklist to a worker. ill commit the changed when git.friendi.ca is up again :)

Commit:
Use worker for CSV import to avoid timeouts

Commit:
Fix code style (binary_operator_spaces, line endings)

Commit:
Fix code style (binary_operator_spaces, line endings)

Commit:
Update messages.po

Commit:
Update messages.po again

Raroun commented:
Since the contact blocking is now done via a worker, we only show how much contact blockings have been queued:

<img width="1024" height="576" alt="image" src="https://github.com/user-attachments/assets/b7fd69ff-ee25-4663-8bc3-13ce93910104" />

Code-Quality / PHP Stan is failing beacuse of:

```
  Line   tesseract/tesseract.php                           
 ------ -------------------------------------------------- 
  94     Call to an undefined method                       
         thiagoalessio\TesseractOCR\TesseractOCR::lang().  
         🪪  method.notFound                  
```             

Im pretty sure i didn´t touch the addons, so im not sure what to do here. Any help would be appreciated :)

annando commented on src/Worker/Contact/BlockFromCSV.php:
```diff
@@ -0,0 +1,55 @@
+use Friendica\Model\Contact;
+
+/**
+ * Block a contact from a CSV import
```
This worker is used to block some contact. It is not limited to a CSV import but could be called from somewhere else as well. So I think that you should give it a more general name.

Raroun replied:
This is a good Point, i will rename the Worker to "BlockByUrl".

Commit:
Rename Worker BlockFromCSV to BlockByUrl for broader use

annando commented on src/Module/Moderation/Blocklist/Contact.php:
```diff
@@ -36,6 +36,49 @@ public function __construct(Database $database, Page $page, AppHelper $appHelper
+
+			if (!empty($contact['addr']) && strpos($contact['addr'], '@') !== false) {
+				$identifier = $contact['addr'];
+			} elseif (!empty($contact['alias']) && !str_starts_with($contact['alias'], 'did:')) {
```
What is this check for `did:` for?

Raroun replied:
The check filters out Bluesky DIDs. For Bluesky contacts, the alias field can contain (or mostly did in my tests) a DID (e.g. did:plc:abc1234), which wouldn't be useful as an identifier in the CSV. Instead, for DID-based contacts, I combined the readable name with the URL (e.g. Raroun (did:plc:xyz123)).
Initially I had just the DID in the CSV, but I thought it would be better to have a human-readable identifier.

annando replied:
But to check for Bluesky contacts, you could simply use the `network` value of the contact.

annando commented on view/templates/moderation/blocklist/contact.tpl:
```diff
@@ -17,6 +17,13 @@
+	<h3>{{$h_newblock}}</h3>
+	<p>
+		<a href="{{$baseurl}}/moderation/blocklist/contact/import" class="btn btn-primary">{{$import_csv}}</a>
+		<a href="{{$baseurl}}/moderation/blocklist/contact?export=1" class="btn btn-default">{{$export_csv}}</a>
```
Why have you chosen this slightly weird url?

Raroun replied:
I chose the query parameter because i thought the export doesn't need a separate view – it just outputs a csv file. The import has its own template, so a separate route made more sense there. But I'm happy to change it to /export  if you prefer.

annando replied:
If possible, I would prefer that. It just looks nicer.

annando reviewed (commented):
Please have a look at my comments.

annando reviewed (changes requested):
Please change the path. Also you have to rebuild the `messages.po` due to conflict.

Commit:
Use /export path instead of ?export=1 query parameter

Commit:
Add missing export route

Commit:
Update messages.po

Raroun commented:
> Please change the path. Also you have to rebuild the `messages.po` due to conflict.

i changed the path and updated the messages.po

<img width="920" height="1292" alt="image" src="https://github.com/user-attachments/assets/9bcefd50-16ae-458f-8a29-1fc8727ba1cc" />

annando commented on src/Module/Moderation/Blocklist/Contact/Export.php:
```diff
@@ -0,0 +1,55 @@
+	{
+		$this->checkModerationAccess();
+
+		$condition = ['uid' => 0, 'blocked' => true];
```
Please change this to `$condition = ['uid' => 0, 'blocked' => true, 'network' => Protocol::FEDERATED];`  to only address contacts in the federated networks.

Raroun replied:
Will do it tomorrow as im busy today.
Thank you for the suggestion!

Raroun replied:
i changed it to only check for federated contacts and removed the DID handling. Thank you!

annando commented:
Please have got a look at my last comment. If that is done, you can omit the check for `did`.

Commit:
Limit export to federated protocols, remove DID handling

Raroun committed:
Merge branch '2025.07-rc' into contact-blocklist-csv
